proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,313 @@
+risky to aggressively reduce block production, so we are not proceeding with
+this proposal for now.
+
+## Detailed Design
+
+The new protocol tries to make all restarting validators get the same data
+blocks and the same set of last votes, then they will almost certainly make the 
+same decision on the canonical fork and proceed.
+
+A new command line arg will be added. When the cluster is in need
+of a restart, we assume validators holding at least `RESTART_STAKE_THRESHOLD`
+percentage of stakes will restart with this arg. Then the following steps
+will happen:
+
+1. The operator restarts the validator with a new command-line argument to
+cause it to enter the silent repair phase at boot, where it will not make new 
+blocks or vote. The validator propagates its local voted fork
+information to all other validators in restart.
+
+2. While counting local vote information from all others in restart, the
+validator repairs all blocks which could potentially have been optimistically
+confirmed.
+
+3. After repair is complete, the validator counts votes on each fork and
+sends out local heaviest fork.
+
+4. Each validator counts if enough nodes can agree on one block (same slot and
+hash) to restart from:
+
+   1. If yes, proceed and restart
+
+   2. If no, print out what it thinks is wrong, halt and wait for human
+
+See each step explained in details below.
+
+### 1. Gossip last vote before the restart and ancestors on that fork
+
+The main goal of this step is to propagate the locally selected fork to all
+others in restart.
+
+We use a new Gossip message `LastVotedForkSlots`, its fields are:
+
+* `last_voted_slot`: `u64` the slot last voted, this also serves as last_slot
+for the bit vector.
+* `last_voted_hash`: `Hash` the bank hash of the slot last voted slot.
+* `ancestors`: `BitVec<u8>` compressed bit vector representing the slots on
+sender's last voted fork. the most significant bit is always
+`last_voted_slot`, least significant bit is `last_voted_slot-81000`.
+
+The number of ancestor slots sent is hard coded at 81000, because that's
+400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9 
+hours. If a validator restarts after 9 hours past the outage, it cannot join 
+the restart this way. If enough validators failed to restart within 9 hours, 
+then fallback to the manual, interactive cluster restart method.
+
+When a validator enters restart, it uses silent repair shred version to avoid
+interfering with those outside the restart. There is slight chance that 
+the silent repair shred version would collide with the shred version after the
+silent repair phase, but even if this rare case occurred, we plan to flush the 
+CRDS table on successful restart, so gossip messages used in restart will be 
+removed.
+
+To be extra cautious, we will also filter out `LastVotedForkSlots` and
+`HeaviestFork` in gossip if a validator is not in silent repair phase.
```

Comment:
Any references to protocol-specific terminology should be in quotes, "silent repair phase."

Reply:
Changed